flytekit/types/flyte_directory.py

```diff
@@ -0,0 +1,223 @@
+    IDL literal/type of Multipart Blob. Please see the FlyteDirectory comments for additional information.
+    """
+    def __init__(self):
+        super().__init__(name="FlyteDirectory", t=FlyteDirectory)
+
+    @staticmethod
+    def get_format(t: typing.Type[FlyteDirectory]) -> str:
+        return t.extension()
+
+    @staticmethod
+    def _blob_type(format: str) -> _core_types.BlobType:
+        return _core_types.BlobType(format=format, dimensionality=_core_types.BlobType.BlobDimensionality.MULTIPART)
+
+    def get_literal_type(self, t: typing.Type[FlyteDirectory]) -> LiteralType:
+        return _type_models.LiteralType(blob=self._blob_type(format=FlyteDirToMultipartBlobTransformer.get_format(t)))
+
+    def to_literal(
+        self,
+        ctx: FlyteContext,
+        python_val: FlyteDirectory,
+        python_type: typing.Type[FlyteDirectory],
+        expected: LiteralType,
+    ) -> Literal:
+
+        remote_directory = None
+        should_upload = True
+
+        # There are two kinds of literals we handle, either an actual FlyteDirectory, or a string path to a directory.
+        # Handle the FlyteDirectory case
+        if isinstance(python_val, FlyteDirectory):
+            source_path = python_val.path
+            if python_val.remote_directory is False:
+                # If the user specified the remote_path to be False, that means no matter what, do not upload
+                should_upload = False
+            else:
+                # Otherwise, if not an "" use the user-specified remote path instead of the random one
+                remote_directory = python_val.remote_directory or None
+
+        # Handle the string case
+        else:
+            if not (isinstance(python_val, os.PathLike) or isinstance(python_val, str)):
+                raise AssertionError(f"Expected FlyteDirectory or os.PathLike object, received {type(python_val)}")
+
+            source_path = python_val
+            # Only do this check if it's a local directory.
+            if not ctx.file_access.is_remote(source_path):
+                p = Path(source_path)
+                if not p.is_dir():
+                    raise AssertionError(f"Expected a directory. {source_path} is not a directory")
+
+        # For remote values, say s3://some/extant/dir/, we will not upload to Flyte's store (S3/GCS)
+        # and just return a literal with a uri equal to the path given
+        if ctx.file_access.is_remote(source_path) or not should_upload:
+            meta = BlobMetadata(type=self._blob_type(format=self.get_format(python_type)))
+            return Literal(scalar=Scalar(blob=Blob(metadata=meta, uri=source_path)))
+
+        # For local paths, we will upload to the Flyte store (note that for local execution, the remote store is just
+        # a subfolder), unless remote_path=False was given
+        else:
+            if remote_directory is None:
+                remote_directory = ctx.file_access.get_random_remote_directory()
```

Comment:
is get_random_remote_directory deterministic? should you save it once you assign it so to_literal() returns the same value?

Reply:
It is random.  I'm gonna think about that as part of another PR though if that's okay...